File: engines/contract_number.py
"""Contract Number Engine — sequential numbers like ZZ-BRAND-2026-0001.

State persisted in output/sequence.json. Never reused, never reset automatically.
"""
import json
import os
import re
from datetime import datetime
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SEQUENCE_FILE = ROOT / 'output' / 'sequence.json'


def _read_sequence() -> dict:
    SEQUENCE_FILE.parent.mkdir(parents=True, exist_ok=True)
    if not SEQUENCE_FILE.exists():
        return {}
    try:
        return json.loads(SEQUENCE_FILE.read_text(encoding='utf-8'))
    except (json.JSONDecodeError, OSError):
        return {}


def _write_sequence(store: dict) -> None:
    SEQUENCE_FILE.write_text(json.dumps(store, indent=2), encoding='utf-8')


def next_contract_number(contract_type: str) -> str:
    """contract_type: 'BRAND' or 'EMP'. Returns e.g. 'ZZ-BRAND-2026-0001'."""
    year = str(datetime.now().year)
    prefix = os.environ.get('CONTRACT_PREFIX', 'ZZ')

    store = _read_sequence()
    store.setdefault(contract_type, {})
    store[contract_type].setdefault(year, 0)
    store[contract_type][year] += 1
    seq = str(store[contract_type][year]).zfill(4)

    _write_sequence(store)
    return f'{prefix}-{contract_type}-{year}-{seq}'


def build_filename(contract_no: str, party_name: str, ext: str) -> str:
    """e.g. ZZ-BRAND-2026-0001_NIKE_INDIA.docx"""
    slug = re.sub(r'[^A-Z0-9]+', '_', party_name.upper()).strip('_')
    return f'{contract_no}_{slug}.{ext}'

File: web/lib/contractNumber.ts
/**
 * Contract Number Engine — faithful TypeScript port of engines/contract_number.py
 *
 * Reads/writes output/sequence.json (one directory up from web/).
 * Never resets, never re-uses numbers.
 */
import fs from 'fs';
import path from 'path';
import { writableDir } from './paths';

const OUTPUT_DIR = writableDir('output');
const SEQUENCE_FILE = path.join(OUTPUT_DIR, 'sequence.json');

function readSequence(): Record<string, Record<string, number>> {
  try {
    if (!fs.existsSync(SEQUENCE_FILE)) return {};
    return JSON.parse(fs.readFileSync(SEQUENCE_FILE, 'utf-8'));
  } catch {
    return {};
  }
}

function writeSequence(data: Record<string, Record<string, number>>): void {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  fs.writeFileSync(SEQUENCE_FILE, JSON.stringify(data, null, 2), 'utf-8');
}

/** Returns e.g. "ZZ-BRAND-2026-0001" or "ZZ-LOR-2026-0001" or "BCPL/NO/105" */
export function nextContractNumber(type: 'BRAND' | 'EMP' | 'CERT' | 'LOR' | 'PI'): string {
  const year = String(new Date().getFullYear());
  const prefix = process.env.CONTRACT_PREFIX ?? 'ZZ';

  const store = readSequence();
  if (!store[type]) store[type] = {};

  if (type === 'PI') {
    return commitPiNumber();
  }

  if (!store[type][year]) store[type][year] = 0;
  store[type][year] += 1;

  const seq = String(store[type][year]).padStart(4, '0');
  writeSequence(store);
  return `${prefix}-${type}-${year}-${seq}`;
}

export function peekNextPiNumber(): number {
  const store = readSequence();
  if (!store['PI']) store['PI'] = {};
  if (store['PI']['all'] === undefined) {
    return Number(process.env.PI_STARTING_NUMBER) || 105;
  }
  return store['PI']['all'] + 1;
}

export function commitPiNumber(manualNumber?: number): string {
  const store = readSequence();
  if (!store['PI']) store['PI'] = {};
  if (store['PI']['all'] === undefined) {
    const starting = Number(process.env.PI_STARTING_NUMBER) || 105;
    store['PI']['all'] = starting - 1;
  }

  if (manualNumber === undefined) {
    store['PI']['all'] += 1;
  } else {
    store['PI']['all'] = Math.max(store['PI']['all'], manualNumber);
  }

  const seqNum = manualNumber !== undefined ? manualNumber : store['PI']['all'];
  writeSequence(store);
  return `BCPL/NO/${seqNum}`;
}

/** e.g. ("ZZ-BRAND-2026-0001", "Nike India", "docx") → "ZZ-BRAND-2026-0001_NIKE_INDIA.docx" */
export function buildFilename(contractNo: string, partyName: string, ext: string): string {
  const cleanContractNo = contractNo.replace(/\//g, '-');
  const slug = partyName.toUpperCase().replace(/[^A-Z0-9]+/g, '_').replace(/^_+|_+$/g, '');
  return `${cleanContractNo}_${slug}.${ext}`;
}
